File: Python/ByRound/1795/1795_A_Two_Towers.py
# Problem: CF 1795 A - Two Towers
# https://codeforces.com/contest/1795/problem/A

"""
Problem: Two Towers
Algorithm/Technique: Simulation with string manipulation and counting
Time Complexity: O(n + m) per test case, where n and m are lengths of the two strings
Space Complexity: O(n + m) due to string concatenation and reversal

This problem involves determining whether we can make two towers "beautiful" by moving 
blocks between them. A tower is beautiful if no two adjacent blocks have the same color.
We simulate the process by concatenating the two strings (reversing the second), then
counting consecutive same-color blocks to check if it's possible to arrange the towers
into a beautiful configuration.
"""

for a in range(int(input())):  # Read number of test cases
    l = input()  # Read n and m (unused in logic but required for input parsing)
    s1 = input()  # First tower string
    s2 = input()[::-1]  # Reverse the second tower string to simulate top block movement
    s = s1 + s2  # Concatenate the two towers
    n = 0  # Counter for consecutive same characters
    for i in range(len(s) - 1):  # Iterate through concatenated string
        if s[i] == s[i + 1]:  # If adjacent characters are same
            n += 1  # Increment counter
    if n < 2:  # If less than two consecutive same characters
        print("YES")  # It's possible to make it beautiful
    else:
        print("NO")  # Otherwise not possible


# https://github.com/VaHiX/codeForces/

File: CPP/ByTier/A/1795_A_Two_Towers.cpp
// Problem: CF 1795 A - Two Towers
// https://codeforces.com/contest/1795/problem/A

/*
 * Problem: A. Two Towers
 * Purpose: Determine if two towers of blocks (represented by strings of 'B' and 'R') 
 *          can be made "beautiful" by moving top blocks between towers, where a 
 *          beautiful configuration means no two adjacent blocks have the same color.
 *
 * Algorithm:
 * 1. For each test case, reverse the second string and append it to the first.
 * 2. Check if there exists any consecutive identical characters in the combined string.
 *    If yes, return "NO"; otherwise, return "YES".
 *
 * Time Complexity: O(n + m) per test case, where n and m are lengths of the two strings.
 * Space Complexity: O(n + m) for the combined string storage.
 */

#include <algorithm>
#include <iostream>
#include <string>

int main() {
  std::ios_base::sync_with_stdio(false);
  long t;
  std::cin >> t;
  while (t--) {
    long n, m;
    std::cin >> n >> m;
    std::string a, b;
    std::cin >> a >> b;
    std::reverse(b.begin(), b.end()); // Reverse second tower and append to first
    a += b; // Combine the two towers (reversed second one appended)
    bool first(false), possible(true);
    for (long p = 1; possible && p < a.size(); p++) {
      bool same = (a[p] == a[p - 1]); // Check if adjacent characters are same
      if (first && same) { // If we've seen a pair of same chars before and current is same
        possible = false;  // Then it's impossible to make it beautiful
        break;
      } else if (same) {
        first = true; // Mark that we have encountered a same character pair
      }
    }
    std::cout << (possible ? "YES" : "NO") << std::endl;
  }
}


// https://github.com/VaHiX/codeForces/